staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go

```diff
@@ -73,7 +75,19 @@ func withWaitGroup(handler http.Handler, longRunning apirequest.LongRunningReque
 			return
 		}
 
+		if terminator != nil && longRunningRequest {
+			newCtx, cancelFn := context.WithCancel(ctx)
```

Comment:
How about API like:

```
ctx, cancelFn := terminator.Attach(ctx)
defer cancelFn()
```

the cancelFn will both remove request from the internal state and cancelFn for the WithCancel.